### ISSUE TITLE
Support deleting stale AnalysisRuns not associated to Rollout

### ISSUE THREAD

**Issue by kevinqian-db** [argoproj/argo-rollouts#3285]
# Summary

Right now Argo rollout supports metrics retention and max history retention for rollouts. However, it seems that it does not support auto-cleanup for one-off AnalysisRuns. In our case we noticed that some of our clusters have thousands of stale ARs, and they cause the metrics server >30s to respond.

# Use Cases

When one-off AnalysisRuns are created for non-Argo controlled resources (e.g. StatefulSets).

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritize the issues with the most 👍.

**Comment by jessesuen** [comment]
One consideration is that because runs can be synthesized from multiple templates, we would need to determine what to do if ttl disagreed. I guess we would need to take the largest value.

**Comment by zachaller** [comment]
So I like the proposal I would like to define exactly how `AnalysisRun.Status.CompletedAt` is determined but overall the idea sounds good.

I think that `(Optional) Add the following to AnalysisTemplate manifest:` might be slightly hard to define because analysis templates get merged together, we could maybe do some heuristic but I am also ok with it just living on the AnalysisRun itself, especially as a first pass.

**Comment by kevinqian-db** [comment]
> I think that (Optional) Add the following to AnalysisTemplate manifest: might be slightly hard to define because analysis templates get merged together, we could maybe do some heuristic but I am also ok with it just being on the AnalysisRun itself, especially as a first pass.

Sounds good. For our immediate usage I don't think we need to support is for now.

> So I like the proposal I would like to define exactly how AnalysisRun.Status.CompletedAt is determined but overall the idea sounds good.

I believe we can make `CompletedAt` as the `now` timestamp of the AnalysisRun reconciliation pass. Right now for `StartedAt` we populate it with `now` during `assessRunStatus` at https://github.com/argoproj/argo-rollouts/blob/99ae20c0ffdfd2543856561fd528fcd51b4fb0a5/analysis/analysis.go#L428-L431

I think for `CompletedAt` we can similarly take the `now` value at https://github.com/argoproj/argo-rollouts/blob/99ae20c0ffdfd2543856561fd528fcd51b4fb0a5/analysis/analysis.go#L105-L112 of when the new phase != existing phase while new phase being a Complete phase.

**Comment by zachaller** [comment]
Yup that makes since I did not know off the top of my head if there was an actual completed phase and that that phase was also set on things like errors or failures etc, if that phase is not set on errors or failures do we still want to GC etc?

**Comment by kevinqian-db** [comment]
> Yup that makes since I did not know off the top of my head if there was an actual completed phase and that that phase was also set on things like errors or failures etc, if that phase is not set on errors or failures do we still want to GC? Just things to double check and confirm etc.

According to https://github.com/argoproj/argo-rollouts/blob/99ae20c0ffdfd2543856561fd528fcd51b4fb0a5/pkg/apis/rollouts/v1alpha1/analysis_types.go#L198-L205

All these phases should be considered complete, so IMO it makes sense to just the value to all these 4 types of phases.

**Comment by zachaller** [comment]
Yea agree, I am open to a PR the implements option 2. along with the addition of the `CompletedAt` status field for use in determining when to GC based on the TTL settings.

**Comment by kevinqian-db** [comment]
> I think an acceptable and unsurprising heuristic is to take the max of all TTLs from all Templates, for the synthesized run.

Will definitely need to document this well to avoid surprises, since some might expect the min instead of max (in the sense of "overriding"). Also I think we can implement this on template in a separate PR (I believe measurementRetention is also similarly separated into 3 PRs)
